bundles/org.openhab.binding.mpd/src/main/java/org/openhab/binding/mpd/internal/handler/MPDHandler.java

```diff
@@ -0,0 +1,289 @@
+ */
+package org.openhab.binding.mpd.internal.handler;
+
+import static org.openhab.binding.mpd.internal.MPDBindingConstants.*;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.IncreaseDecreaseType;
+import org.eclipse.smarthome.core.library.types.NextPreviousType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.PercentType;
+import org.eclipse.smarthome.core.library.types.PlayPauseType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerService;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.openhab.binding.mpd.internal.MPDConfiguration;
+import org.openhab.binding.mpd.internal.action.MPDActions;
+import org.openhab.binding.mpd.internal.protocol.MPDConnection;
+import org.openhab.binding.mpd.internal.protocol.MPDSong;
+import org.openhab.binding.mpd.internal.protocol.MPDStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link MPDHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Stefan Röllin - Initial contribution
+ */
+@NonNullByDefault
+public class MPDHandler extends BaseThingHandler implements MPDEventListener {
+
+    private final Logger logger = LoggerFactory.getLogger(MPDHandler.class);
+
+    private Map<String, @Nullable State> stateMap = Collections.synchronizedMap(new HashMap<String, @Nullable State>());
+
+    private final MPDConnection connection;
+    private int volume = 0;
+
+    private @Nullable ScheduledFuture<?> futureUpdateStatus;
+    private @Nullable ScheduledFuture<?> futureUpdateCurrentSong;
+
+    public MPDHandler(Thing thing) {
+        super(thing);
+        connection = new MPDConnection(this);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            handleCommandRefresh(channelUID.getId());
+        } else {
+            handlePlayerCommand(channelUID.getId(), command);
+        }
+    }
+
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
```

Comment:
After the thing status is set to unknown, the framework starts to work with this Thing: `handleCommand()` and `dispose()` can be invoked (concurrently). You might want to move it to the end of this method.

Reply:
In theory, the thread could update the status before updateStatus from the main thread is called. Therefore I would prefer to leave it before the start of the thread.